File: java/byps/src/com/wilutions/byps/BAsyncResultReceiveMethod.java
package com.wilutions.byps;

/**
 * This class wraps a BMethodResult object into an BAsyncResult.
 * Objects of this class are used in generated stubs.
 *
 * @param <T> The method's result type.
 */
public class BAsyncResultReceiveMethod<T> implements BAsyncResult<BMethodResult<T>> {
	
	private final BAsyncResult<T> innerResult;
	
	public BAsyncResultReceiveMethod(BAsyncResult<T> innerResult) {
		this.innerResult = innerResult;
	}
	
	@Override
	public void setAsyncResult(BMethodResult<T> methodResult, Throwable e)  {
		T result = methodResult != null ? methodResult.result : null;
		innerResult.setAsyncResult(result, e);
	}
	
}

File: csharp/byps/BAsyncResultReceiveMethod.cs
/* USE THIS FILE ACCORDING TO THE COPYRIGHT RULES IN LICENSE.TXT WHICH IS PART OF THE SOURCE CODE PACKAGE */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace byps
{
    public class BAsyncResultReceiveMethod<T> : BAsyncResultIF<BMethodResult<T>>
    {
        private readonly BAsyncResultIF<T> innerResult;
        private readonly BAsyncResult<T> fnct;

        public BAsyncResultReceiveMethod(BAsyncResultIF<T> outerResult)
        {
            this.innerResult = outerResult;
        }

        public BAsyncResultReceiveMethod(BAsyncResult<T> fnct)
        {
            this.fnct = fnct;
        }

        public void setAsyncResult(BMethodResult<T> methodResult, Exception e)
        {
            T result = methodResult != null ? methodResult.resultValue : default(T);
            if (innerResult != null)
            {
                innerResult.setAsyncResult(result, e);
            }
            else
            {
                fnct(result, e);
            }
        }
    }
}
